### PR TITLE
update systemlog indexer and evm transfer api

### PR DESCRIPTION
**Pull request by Frankonly**
Giving more detail status of error for evm transfer APIs.

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -751,11 +752,14 @@ func (api *Server) GetEvmTransfersByActionHash(ctx context.Context, in *iotexapi
 
 	actHash, err := hash.HexStringToHash256(in.ActionHash)
 	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
 	transfers, err := api.systemLogIndexer.GetEvmTransfersByActionHash(actHash)
 	if err != nil {
+		if strings.Contains(err.Error(), "key = ") && strings.Contains(err.Error(), "doesn't exist") {
```
Comment (lizhefeng) [inline api/api.go:760]: This check is too hacky. Can we use errors.Cause() to check the root cause? If you don't wrap the error in systemLogIndexer.GetEvmTransfersByActionHash(actHash), please wrap it with a ErrNotExist so you can check here.

File: api/api.go [hunk 2]
```diff
@@ -768,8 +772,18 @@ func (api *Server) GetEvmTransfersByBlockHeight(ctx context.Context, in *iotexap
 		return nil, status.Error(codes.Unavailable, "evm transfer index not supported")
 	}
 
+	if in.BlockHeight < 1 {
+		return nil, status.Errorf(codes.OutOfRange, "invalid block height = %d", in.BlockHeight)
+	}
+
 	transfers, err := api.systemLogIndexer.GetEvmTransfersByBlockHeight(in.BlockHeight)
 	if err != nil {
+		if strings.Contains(err.Error(), "key = ") && strings.Contains(err.Error(), "doesn't exist") {
```
Comment (lizhefeng) [inline api/api.go:781]: Same here.

File: api/api.go [hunk 3]
```diff
@@ -768,8 +772,18 @@ func (api *Server) GetEvmTransfersByBlockHeight(ctx context.Context, in *iotexap
...
+			return nil, status.Error(codes.NotFound, "no such block with evm transfer")
+		}
+		if strings.Contains(err.Error(), systemlog.ErrHighBlockHeight.Error()) {
```
Comment (lizhefeng) [inline api/api.go:784]: Same here.

File: api/api.go [hunk 4]
```diff
@@ -768,8 +772,18 @@ func (api *Server) GetEvmTransfersByBlockHeight(ctx context.Context, in *iotexap
 		return nil, status.Error(codes.Unavailable, "evm transfer index not supported")
 	}
 
+	if in.BlockHeight < 1 {
+		return nil, status.Errorf(codes.OutOfRange, "invalid block height = %d", in.BlockHeight)
```
Comment (lizhefeng) [inline api/api.go:776]: To be consistent with other existing APIs, use codes.InvalidArgument here as well.

File: api/api.go [hunk 5]
```diff
@@ -768,8 +772,18 @@ func (api *Server) GetEvmTransfersByBlockHeight(ctx context.Context, in *iotexap
...
+	}
+
 	transfers, err := api.systemLogIndexer.GetEvmTransfersByBlockHeight(in.BlockHeight)
 	if err != nil {
+		if strings.Contains(err.Error(), "key = ") && strings.Contains(err.Error(), "doesn't exist") {
+			return nil, status.Error(codes.NotFound, "no such block with evm transfer")
+		}
+		if strings.Contains(err.Error(), systemlog.ErrHighBlockHeight.Error()) {
+			return nil, status.Errorf(codes.OutOfRange, "height = %d is higher than current height", in.BlockHeight)
```
Comment (lizhefeng) [inline api/api.go:785]: Ditto.

File: systemlog/systemlog.go [hunk 6]
```diff
@@ -233,6 +242,13 @@ func (x *Indexer) GetEvmTransfersByActionHash(actionHash hash.Hash256) (*iotexty
 
 // GetEvmTransfersByBlockHeight queries evm transfers by block height
 func (x *Indexer) GetEvmTransfersByBlockHeight(blockHeight uint64) (*iotextypes.BlockEvmTransfer, error) {
+	tipHeight, err := x.tipHeight()
+	if err != nil {
+		return nil, err
+	}
+	if tipHeight < blockHeight {
+		return nil, ErrHighBlockHeight
```
Comment (lizhefeng) [inline systemlog/systemlog.go:250]: ErrHighBlockHeight?

File: api/api.go (hunk 4 above)
Comment (Frankonly) [reply api/api.go:776]: ok

File: api/api.go [hunk 7]
```diff
@@ -771,16 +771,16 @@ func (api *Server) GetEvmTransfersByBlockHeight(ctx context.Context, in *iotexap
...
 
 	transfers, err := api.systemLogIndexer.GetEvmTransfersByBlockHeight(in.BlockHeight)
 	if err != nil {
-		if strings.Contains(err.Error(), "key = ") && strings.Contains(err.Error(), "doesn't exist") {
+		if err == systemlog.ErrNotFound {
 			return nil, status.Error(codes.NotFound, "no such block with evm transfer")
 		}
-		if strings.Contains(err.Error(), systemlog.ErrHighBlockHeight.Error()) {
-			return nil, status.Errorf(codes.OutOfRange, "height = %d is higher than current height", in.BlockHeight)
+		if strings.Contains(err.Error(), systemlog.ErrHeightNotReached.Error()) {
```
Comment (lizhefeng) [inline api/api.go:782]: Use errors.Cause().

File: systemlog/systemlog.go [hunk 8]
```diff
@@ -215,6 +224,9 @@ func (x *Indexer) DeleteTipBlock(blk *block.Block) error {
 func (x *Indexer) GetEvmTransfersByActionHash(actionHash hash.Hash256) (*iotextypes.ActionEvmTransfer, error) {
 	data, err := x.kvStore.Get(evmTransferNS, actionKey(actionHash))
 	if err != nil {
+		if errors.Cause(err) == db.ErrNotExist {
```
Comment (lizhefeng) [inline systemlog/systemlog.go:227]: This is nit. db.ErrNotExist should be caught by the upper level.

File: systemlog/systemlog.go [hunk 9]
```diff
@@ -233,8 +245,18 @@ func (x *Indexer) GetEvmTransfersByActionHash(actionHash hash.Hash256) (*iotexty
...
+		return nil, err
+	}
+	if tipHeight < blockHeight {
+		return nil, ErrHeightNotReached
+	}
 	data, err := x.kvStore.Get(evmTransferNS, blockKey(blockHeight))
 	if err != nil {
+		if errors.Cause(err) != db.ErrNotExist {
+			return nil, err
+		}
```
Comment (lizhefeng) [inline systemlog/systemlog.go:259]: ditto.

**Pull request merged**